concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+Here's a few (non-exhaustive) examples in a handful of languages:
+
+### enums in the wild
+
+In Java, the `enum` keyword maps directly the category of [enum](#enums) we've defined here.
+
+In Go, *there is no direct support for enums*.
+However, it's often implemented in practice by a typedef of an int, and a table of named constants.
+
+In Rust, there's an `enum` keyword, and it may be used to implement [enum](#enums) as we've defined it here.
+However, if you're coming from a Rust background, and trying to understand this document and compare to other languages,
+be cautious here: the Rust `enum` keyword *also* provides what we call [sum types](#sum-types).
+
+#### booleans
+
+In many languages, the concept of `boolean` can be understood as a particular instance of an enum --
+it's just an enum with two values!
+Sure, it may be implemented as 'primitive' in many languages... but clearly,
+for cardinality counting, we can see a boolean as exactly "|2|".
+
+### sum types in the wild
+
+In Java, *there is no direct support for sum types*.
+The nearest thing possible is to create an interface,
+and promise that only a known number of types will implement it.
+
+In Go, *there is no direct support for sum types*.
+The nearest thing possible is to create an interface with an unexported method,
+and make sure only a known number of types in the same package implement it.
+(See https://github.com/BurntSushi/go-sumtype for tooling to help support this;
+the readme also contains good documentation of the concept!)
+
+In Rust, the `enum` keyword (confusingly) also creates sum types.
+Rust's sum types are at least properly closed (unlike Java's and Go's).
+
+#### maybe types
+
+In *many* languages, you may find a type in the standard library called "Maybe",
+or equally frequently, "Optional", or, "Either".
+These are a sum type!
+They're typically explained as "either **none**, or **some** of a thing":
+so, they take the cardinality of a thing and simply "plus one" to it,
+and colloquially call that additional state the "none" state.
+
+For some `Maybe<T>`, the cardinality is "1+|T|".
+
+(This is also a great example for how we can use cardinality counting methods
+to create clear comparisons between differently named but functionally
+identical concepts between programming language communities!)
+
+### product types in the wild
+
+In Java, `class` creates a product type.
+
+In Go, `struct` creates a product type.
+
+In Rust, `struct` creates a product type.
+
+### interfaces in the wild
+
+In Java, the `interface` keyword introduces an interface.
+These interfaces are [*explicit*](#explicit-versus-structural-interfaces) -- a class must use the `implements` keyword to explicitly state that it can satisfy that interface.
+(That the class has sufficient methods to match the interface is then also checked at compile time -- but having the methods is not necessary rather than sufficient for the "implements" relationship to be considered true).
```

Comment:
```suggestion
(That the class has sufficient methods to match the interface is then also checked at compile time -- but having the methods is necessary rather than sufficient for the "implements" relationship to be considered true).
```